runtime/parachains/src/disputes/slashing.rs

```diff
@@ -0,0 +1,905 @@
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Dispute slashing pallet.
+//!
+//! The implementation relies on the `offences` pallet and
+//! looks like a hybrid of `im-online` and `grandpa` equivocation handlers.
+//! Meaning, we submit an `offence` for the concluded disputes about
+//! the current session candidate directly from the runtime.
+//! If, however, the dispute is about a past session, we record pending
+//! slashes on chain, without `FullIdentification` of the offenders.
+//! Later on, a block producer can submit an unsigned transaction with
+//! `KeyOwnershipProof` of an offender and submit it to the runtime
+//! to produce an offence.
+//! The reason for this separation is that we do not want to rely on
+//! `FullIdentification` to be available on chain for the past sessions
+//!  because it is quite heavy.
+
+use crate::{
+	disputes,
+	initializer::ValidatorSetCount,
+	session_info::{AccountId, IdentificationTuple},
+};
+use frame_support::{
+	traits::{Get, KeyOwnerProofSystem, ValidatorSet, ValidatorSetWithIdentification},
+	weights::{Pays, Weight},
+};
+use parity_scale_codec::{Decode, Encode};
+use primitives::v2::{CandidateHash, SessionIndex, ValidatorId, ValidatorIndex};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::Convert,
+	transaction_validity::{
+		InvalidTransaction, TransactionPriority, TransactionSource, TransactionValidity,
+		TransactionValidityError, ValidTransaction,
+	},
+	DispatchResult, KeyTypeId, Perbill, RuntimeDebug,
+};
+use sp_session::{GetSessionNumber, GetValidatorCount};
+use sp_staking::offence::{DisableStrategy, Kind, Offence, OffenceError, ReportOffence};
+use sp_std::{
+	collections::btree_map::{BTreeMap, Entry},
+	prelude::*,
+};
+
+const LOG_TARGET: &str = "runtime::slashing";
+
+#[cfg(feature = "runtime-benchmarks")]
+pub mod benchmarking;
+
+/// Timeslots should uniquely identify offences
+/// and are used for the offence deduplication.
+#[derive(Eq, PartialEq, Ord, PartialOrd, Clone, Encode, Decode, TypeInfo, RuntimeDebug)]
+pub struct DisputesTimeSlot {
+	// The order of these matters for `derive(Ord)`.
+	session_index: SessionIndex,
+	candidate_hash: CandidateHash,
+}
+
+impl DisputesTimeSlot {
+	pub fn new(session_index: SessionIndex, candidate_hash: CandidateHash) -> Self {
+		Self { session_index, candidate_hash }
```

Comment:
Do we really need this constructor?

Reply:
Yes :)
Otherwise, I'd have to make the fields public, because it's used in benchmarking as well.

Also I wasn't sure if this comment https://github.com/paritytech/substrate/blob/6401300a2ff7ce013385f2c61fc9a3798bd4272d/primitives/staking/src/offence.rs#L75 still holds (I think it never did).